### PR TITLE
Adds redis_data_info module

### PR DESCRIPTION
**Pull request by paginabianca**
Added:
  - redis_data_info module and suggested 'exists' return flag.
  - module_utils for redis with a base class that handles database connections.
  - inhereited unit tests and added some new ones for the exit flag

##### SUMMARY
Adds `redis_data_info` module. Originally from #3144. Applied suggested changes and added some more unit tests.

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME
`redis_data_info`

### CODE REVIEW

File: plugins/doc_fragments/redis.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      - Specify the password to authenticate with.
+      - Usually not used when target is localhost.
+    type: str
+  validate_certs:
+    description:
+      - Specify whether or not to validate SSL certificates.
+      - This should only be turned off for personally controlled sites or with
+        C(localhost) as target.
+    type: bool
+    default: True
```
Comment (felixfontein) [inline plugins/doc_fragments/redis.py:39]:
```suggestion
    default: true
```

File: plugins/doc_fragments/redis.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  ssl_ca_certs:
+    description:
+      - Path to root certificates file. If not set and C(validate_certs) is
+      - set to I(True), certifi ca-certificates will be used.
```
Comment (felixfontein) [inline plugins/doc_fragments/redis.py:43]:
```suggestion
      - Path to root certificates file. If not set and I(validate_certs) is
      - set to C(true), certifi ca-certificates will be used.
```

File: plugins/doc_fragments/redis.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      - Usually not used when target is localhost.
+    type: str
+  validate_certs:
+    description:
+      - Specify whether or not to validate SSL certificates.
+      - This should only be turned off for personally controlled sites or with
+        C(localhost) as target.
+    type: bool
+    default: True
+  ssl_ca_certs:
```
Comment (felixfontein) [inline plugins/doc_fragments/redis.py:40]:
```suggestion
  ca_certs:
```

File: plugins/module_utils/redis.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            ssl_ca_certs = str(certifi.where())
+
+        try:
+            if validate_certs:
+                return Redis(host=login_host,
+                             port=login_port,
+                             username=login_user,
+                             password=login_password,
+                             ssl_ca_certs=ssl_ca_certs,
+                             ssl=validate_certs,)
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:79]: This looks wrong. `ssl` sets whether to use TLS or not, not whether to validate TLS certificates or not.

You probably want another option which controls whether TLS is used (or not). `validate_certs` and `ca_certs` should only be used when that flag is set.

File: plugins/modules/database/misc/redis_data_info.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: redis_data_info
+short_description: Get value of key in Redis database
+version_added: 3.6.0
+description:
+  - Get value of keys in Redis database
```
Comment (felixfontein) [inline plugins/modules/database/misc/redis_data_info.py:16]:
```suggestion
  - Get value of keys in Redis database.
```

File: plugins/modules/database/misc/redis_data_info.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+author: "Andreas Botzner (@paginabianca)"
+options:
+  key:
+    description:
+      - Database key.
+    type: str
+    required: True
```
Comment (felixfontein) [inline plugins/modules/database/misc/redis_data_info.py:23]:
```suggestion
    required: true
```

File: plugins/modules/database/misc/redis_data_info.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+extends_documentation_fragment:
+  - community.general.redis.documentation
```
Comment (felixfontein) [inline plugins/modules/database/misc/redis_data_info.py:26]:
```suggestion
  - community.general.redis
```

File: plugins/modules/database/misc/redis_data_info.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+    login_host: localhost
+    login_password: supersecret
+    key: foo
+
+- name: Get key foo=bar on redishost with custom ca-cert file
+  community.general.redis_data_info:
+    login_host: redishost
+    login_password: supersecret
+    login_user: somuser
+    validate_certs: yes
```
Comment (felixfontein) [inline plugins/modules/database/misc/redis_data_info.py:45]:
```suggestion
    validate_certs: true
```
We should always use `true` and `false` for YAML booleans, and not the other variants.

File: plugins/modules/database/misc/redis_data_info.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    key: foo
+'''
+
+RETURN = '''
+exists:
+  description: If they key exists in the database.
+  returned: on success
+  type: bool
+value:
+  description: Value key was set to
```
Comment (felixfontein) [inline plugins/modules/database/misc/redis_data_info.py:56]:
```suggestion
  description: Value key was set to.
```

File: plugins/modules/database/misc/redis_data_info.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        msg = 'Failed to get value of key: {0} with exception: {1}'.format(
+            key, str(e))
+        result['msg'] = msg
+        module.fail_json(**result)
+
+    if value is None:
+        msg = 'Key: {0} does not exist in databse'.format(key)
+        result['exists'] = False
+    else:
+        msg = 'Got key: {0} with value: {1}'.format(key, value)
```
Comment (felixfontein) [inline plugins/modules/database/misc/redis_data_info.py:105]: I would avoid replicating the key's value here.
```suggestion
        msg = 'Got key "{0}"'.format(key)
```

File: plugins/modules/database/misc/redis_data_info.py [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    try:
+        value = redis.connection.get(key)
+    except Exception as e:
+        msg = 'Failed to get value of key: {0} with exception: {1}'.format(
+            key, str(e))
+        result['msg'] = msg
+        module.fail_json(**result)
+
+    if value is None:
+        msg = 'Key: {0} does not exist in databse'.format(key)
```
Comment (felixfontein) [inline plugins/modules/database/misc/redis_data_info.py:102]:
```suggestion
        msg = 'Key "{0}" does not exist in database'.format(key)
```

File: plugins/modules/database/misc/redis_data_info.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    redis = RedisAnsible(module)
+
+    key = module.params['key']
+    result = {'changed': False}
+
+    value = None
+    try:
+        value = redis.connection.get(key)
+    except Exception as e:
+        msg = 'Failed to get value of key: {0} with exception: {1}'.format(
```
Comment (felixfontein) [inline plugins/modules/database/misc/redis_data_info.py:96]:
```suggestion
        msg = 'Failed to get value of key "{0}" with exception: {1}'.format(
```

File: plugins/doc_fragments/redis.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    description:
+      - Specify whether or not to validate SSL certificates.
+      - This should only be turned off for personally controlled sites or with
+        C(localhost) as target.
+    type: bool
+    default: true
+  ssl_ca_certs:
+    description:
+      - Path to root certificates file. If not set and C(ssl) is
+      - set to C(True), certifi ca-certificates will be used.
```
Comment (felixfontein) [inline plugins/doc_fragments/redis.py:48]:
```suggestion
      - Path to root certificates file. If not set and I(ssl) is
        set to C(true), certifi ca-certificates will be used.
```

File: plugins/module_utils/redis.py [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+__metaclass__ = type
+
+import traceback
+
+REDIS_IMP_ERR = None
+try:
+    from redis import Redis
+    HAS_REDIS_PACKAGE = True
+except ImportError:
+    REDIS_IMP_ERR = traceback.format_exc()
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:18]: It looks like this is never passed to the user. (Same for the other traceback.)

File: plugins/module_utils/redis.py [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+            ssl_ca_certs = str(certifi.where())
+
+        try:
+            if ssl:
+                return Redis(host=login_host,
+                             port=login_port,
+                             username=login_user,
+                             password=login_password,
+                             ssl_ca_certs=ssl_ca_certs,
+                             ssl_cert_reqs=validate_certs,
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:82]: `ssl_cert_reqs` needs to be a string, not a boolean (https://github.com/andymccurdy/redis-py/blob/master/redis/connection.py#L849-L851). You need to convert the value accordingly.

File: plugins/doc_fragments/redis.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+  login_user:
+    description:
+      - Specify the user to authenticate with.
+    type: str
+  login_password:
+    description:
+      - Specify the password to authenticate with.
+      - Usually not used when target is localhost.
+    type: str
+  ssl:
```
Comment (felixfontein) [inline plugins/doc_fragments/redis.py:33]: I would call this option `tls` or `use_tls`. The protocol hasn't been SSL for a loong time.

Comment (paginabianca) [reply plugins/doc_fragments/redis.py:33]: :+1:  Also renamed `ssl_ca_certs` to `ca_certs` for consistency.

File: plugins/doc_fragments/redis.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+    type: str
+  login_port:
+    description:
+      - Specify the port to connect to.
+    default: 6379
+    type: int
+  login_user:
+    description:
+      - Specify the user to authenticate with.
+      - Requires I(redis >=3.4.0).
```
Comment (felixfontein) [inline plugins/doc_fragments/redis.py:27]:
```suggestion
      - Requires L(redis,https://pypi.org/project/redis/) >= 3.4.0.
```
I(...) is reserved for options. And just writing `redis` can make it confusing.

File: plugins/module_utils/redis.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+    CERTIFI_IMPORT_ERROR = traceback.format_exc()
+    HAS_CERTIFI_PACKAGE = False
+
+
+def fail_imports(module):
+    errors = []
+    traceback = []
+    if not HAS_REDIS_PACKAGE:
+        errors += missing_required_lib('redis') + '\n'
+        traceback += REDIS_IMP_ERR
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:35]:
```suggestion
        traceback.append(REDIS_IMP_ERR)
```

File: plugins/module_utils/redis.py [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+    if not HAS_CERTIFI_PACKAGE:
+        errors += missing_required_lib('certifi')
+        traceback += CERTIFI_IMPORT_ERROR
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:38]:
```suggestion
        traceback.append(CERTIFI_IMPORT_ERROR)
```

File: plugins/module_utils/redis.py [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    if errors:
+        module.fail_json(errors=errors, traceback=traceback)
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:40]:
```suggestion
        module.fail_json(errors=errors, traceback='\n'.join(traceback))
```

File: plugins/module_utils/redis.py [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+except ImportError:
+    CERTIFI_IMPORT_ERROR = traceback.format_exc()
+    HAS_CERTIFI_PACKAGE = False
+
+
+def fail_imports(module):
+    errors = []
+    traceback = []
+    if not HAS_REDIS_PACKAGE:
+        errors += missing_required_lib('redis') + '\n'
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:34]:
```suggestion
        errors.append(missing_required_lib('redis'))
```

File: plugins/module_utils/redis.py [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+        traceback += REDIS_IMP_ERR
+    if not HAS_CERTIFI_PACKAGE:
+        errors += missing_required_lib('certifi')
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:37]:
```suggestion
        errors.append(missing_required_lib('certifi'))
```

File: plugins/module_utils/redis.py [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+        module.fail_json(errors=errors, traceback=traceback)
+
+
+def redis_auth_argument_spec():
+    return dict(
+        login_host=dict(type='str',
+                        default='localhost',),
+        login_user=dict(type='str'),
+        login_password=dict(type='str',
+
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:49]:
```suggestion
```

File: plugins/module_utils/redis.py [hunk 24]
```diff
@@ -0,0 +1,94 @@
...
+        validate_certs = 'required' if self.module.params['validate_certs'] else None
+        ca_certs = self.module.params['ca_certs']
+        if tls and ca_certs is None:
+            ca_certs = str(certifi.where())
+        if tuple(map(int, redis_version.split('.'))) < (3, 4, 0) and login_user is not None:
+            self.module.fail_json(
+                msg='The option `username` in only supported with redis >= 3.4.0.')
+        params = {'host': login_host,
+                  'port': login_port,
+                  'password': login_password,
```
Comment (felixfontein) [inline plugins/module_utils/redis.py:84]: Does `password` work with redis < 3.4.0?

File: plugins/modules/database/misc/redis_data_info.py [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = '''
+---
+module: redis_data_info
+short_description: Get value of key in Redis database
+version_added: 3.6.0
```
Comment (felixfontein) [inline plugins/modules/database/misc/redis_data_info.py:14]:
```suggestion
version_added: 3.7.0
```

File: plugins/module_utils/redis.py (hunk 24 above)
Comment (paginabianca) [reply plugins/module_utils/redis.py:84]: I checked and the `password` kwarg is supported back to 2.4.6 which is the oldest tagged version in the repository.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
Since nobody else seems to want to review this, I'm going to just merge :)

**Comment by felixfontein** [comment]
@paginabianca thanks for contributing this module!

**Comment by paginabianca** [comment]
Awesome thanks for the help along the way :pray:

**Pull request merged**